pkg/cmd/docs-issue-generation/docs_issue_generation.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"log"
+	"net/http"
+	"regexp"
+	"strconv"
+	"time"
+)
+
+type commit struct {
+	sha         string
+	title       string
+	releaseNote string
+}
+
+type pr struct {
+	number      int
+	mergeBranch string
+	commits     []commit
+}
+
+type ghSearch struct {
+	Items []struct {
+		Number      int `json:"number"`
+		PullRequest struct {
+			MergedAt *time.Time `json:"merged_at"`
+		} `json:"pull_request"`
+	} `json:"items"`
+}
+
+type ghPull struct {
+	Base struct {
+		Ref string `json:"ref"`
+	} `json:"base"`
+}
+
+type ghPullCommit struct {
+	Sha    string `json:"sha"`
+	Commit struct {
+		Message string `json:"message"`
+	} `json:"commit"`
+}
+
+func docsIssueGeneration(params parameters) {
+	bsSearch := httpGet("https://api.github.com/search/issues?q=sha:"+params.Sha+"+repo:cockroachdb/cockroach+is:merged", params.Token)
+	prs := prNums(bsSearch)
+	for _, pr := range prs {
+		bsPull := httpGet("https://api.github.com/repos/cockroachdb/cockroach/pulls/"+strconv.Itoa(pr.number), params.Token)
+		pr.mergeBranch = setMergeBranch(bsPull)
+		bsCommit := httpGet("https://api.github.com/repos/cockroachdb/cockroach/pulls/"+strconv.Itoa(pr.number)+"/commits?per_page:250", params.Token)
+		pr.commits = setCommits(bsCommit, pr.number)
+		pr.createDocsIssues(params.Token)
+	}
+}
+
+func httpGet(url string, token string) []byte {
+	client := &http.Client{}
+	req, err1 := http.NewRequest("GET", url, nil)
+	if err1 != nil {
+		log.Fatal(err1)
+	}
+	req.Header.Set("Authorization", "token "+token)
+	res, err2 := client.Do(req)
+	if err2 != nil {
+		log.Fatal(err2)
+	}
+	bs, err3 := io.ReadAll(res.Body)
+	if err3 != nil {
+		log.Fatal(err3)
+	}
+	return bs
+}
+
+func prNums(bs []byte) []pr {
+	search := ghSearch{}
```

Comment:
I think having all your functions take a `[]byte` argument is not what you want to do. Makes it very difficult to tell what each of these functions is meant to do, and I think is going to make your life much harder than you need when the time comes to test this. The caller can convert from `[]byte` to whatever type it needs (you can provide helper functions to do the same if you want). 

At the very least, if your function must take an unstructured array of bytes as an argument, a more informative parameter name than `bs` would help.

Reply:
So `httpGet` returns a byte slice because it's the common ancestor among each of the 3 structs (ghSearch, ghPull, ghPullCommit) I created to house the data for each of my 3 helper functions. Within each helper function, I then unmarshal the byte slice into the struct of data.

I can create 3 more helper functions just to do the unmarshaling. Does that sound in line with your suggestion?

I also plan on extensively commenting the code, but only after I work through these nits, so that I don't accidentally document something that I have to change or remove.

Reply:
> So httpGet returns a byte slice because it's the common ancestor among each of the 3 structs (ghSearch, ghPull, ghPullCommit) I created to house the data for each of my 3 helper functions. Within each helper function, I then unmarshal the byte slice into the struct of data.

I get that's the intention, but it's still possible to achieve code reuse without passing around the raw byte slices to do all your work. You give up a lot of compile-time safety guarantees by doing that.

You can create more helper functions to do the unmarshaling. Reflection is also possible, like:


```
func httpGet(url string, token string, reflect.Type ty) (interface{}, error) {
	client := &http.Client{}
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
      return nil, err
	}
	req.Header.Set("Authorization", "token "+token)
	res, err := client.Do(req)
	if err != nil {
      return nil, err
	}
	bs, err := io.ReadAll(res.Body)
	if err != nil {
      return nil, err
	}
	ret := reflect.New(ty)
  if err := json.Unmarshal(bs, ret); err != nil {
      return nil, err
  }
  return ret, nil
}

res, err := httpGet(url, token, (*ghSearch)(nil))
if err != nil {
    ....
}
search := res.(*ghSearch)
```

(We're not using generics yet so it's not an option.)

Not saying you have to use reflection, having specific helper functions for unmarshaling the 3 struct types you have might be easier, just demonstrating an alternate approach.

BTW: as this code block demonstrates, you can feel free to shadow the name `err`. No need to define many different `errN` variables.

Reply:
Done.